[TEP-0089] - Spire Package

pxp928 opened the pull request:
Signed-off-by: pxp928 <parth.psu@gmail.com>

<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Spire package separated out from https://github.com/tektoncd/pipeline/pull/4759 as requested. It includes the spire interface with a mocked spire for testing. As requested by @afrittoli.

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in
(if there are no user facing changes, use release note "NONE")

```release-note
Spire package that includes the controller and entrypointer API that is used in for [TEP-0089: Non-falsifiable provenance support](https://github.com/tektoncd/community/blob/main/teps/0089-nonfalsifiable-provenance-support.md). This is just the spire package. The phase 1 PR for the TEP is located here -> https://github.com/tektoncd/pipeline/pull/4759 
```

afrittoli commented:
/ok-to-test

afrittoli commented on pkg/spire/controller.go:
```diff
@@ -0,0 +1,276 @@
+}
+
+// NewSpireControllerAPIClient creates a new NewSpireControllerAPIClient for the pipeline controller
+func NewSpireControllerAPIClient(c spireconfig.SpireConfig) ControllerAPIClient {
+	if c.MockSpire {
+		return &MockClient{}
+	}
+	return &spireControllerAPIClient{
+		config: c,
+	}
+}
```
I would prefer to have two different factories here, one the tests and one for the actual usage, instead of embedding a `MockSpire` attribute in the client struct `spireControllerAPIClient`, it would be more consistent with all the other clients used in the controllers, like Tekton API clients, cloudevents client and cache client.

For instance, if you look at the https://github.com/tektoncd/pipeline/tree/main/pkg/reconciler/events/cache package, there are two clients defined there, the [regular one](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/events/cache/cache.go) and the [mock one](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/events/cache/cachefakeclient.go). 

The `knative/pkg` injection machinery makes sure that the correct client (regular or mock) is injected into the context, so that when you [get the client from the context](https://github.com/tektoncd/pipeline/blob/65d3bf3b2d1a828c4598ede08416aab899e2d779/pkg/reconciler/events/cloudevent/cloud_event_controller.go#L143) you get a different one, depending on whether it's unit tests or actual running code.

afrittoli replied:
@mattmoor may want to have a say on this too 🙏

pxp928 replied:
Interesting. I did not know about this. I will update to use that.

pxp928 replied:
Hmm I see. I might need some help setting this up. I am not sure how this would work in this context.

lumjjb replied:
seems do-able, does this look about right? https://github.com/pxp928/pipeline/pull/26/files

afrittoli replied:
Thanks @lumjjb - that's definitely in the right direction! 

The only question I have is whether we could get the config from the context, so that when the client are setup the config is added directly, and we don't need to add it later on. Ideally we should add the client config once, unless we need to support reloading configuration changes at run time. Wdyt?

lumjjb replied:
Thanks @afrittoli ! It's in a weird spot with the [initialization of the TaskRun reconciler](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/taskrun/controller.go#L47), since some of the options come in as CLI arguments. Do you have any recommendations to handle this?

afrittoli commented:
Thanks @pxp928 for moving this to a separate PR!

I'm reviewing the PR, but I find the module structure and test strategy not 100% clear:
In my understanding this is the role of each module is:

- *spire.go*: defines two interfaces, `ControllerAPIClient` and `EntrypointerAPIClient`. (NIT: maybe we call call this *clients*?)
- *controller.go*: implements some of `ControllerAPIClient`, including type, constructor and some of the required method, except for `VerifyTaskRunResults`
- *verify.go*: implements the rest of `ControllerAPIClient`, specifically `VerifyTaskRunResults` (NIT: maybe call this controller_verify?)
-  *entrypointer.go*: implements some of  `EntrypointerAPIClient`, including type, constructor and some of the required method, except for `Sign `
- *sign.go*: implements the rest of `EntrypointerAPIClient`, specifically `VerifyTaskRunResults`, specifically `VerifyTaskRunResults` (NIT: maybe call this controller_verify?)esults` (NIT: maybe call this entrypointer_sign?)
- *spire_mock.go*: defines a mock implementation of both interfaces. This bypasses most of the code in all the actual implementation, but it does invoke some of the actual functions, like `verifyManifest` and `getResultValue`
- *spire_mock_tests.go*: unit tests for *spire_mock.go*. These tests verify the logic in the mocks and some private functions in the *verify* and *sign* modules, but I find this a bit problematic:
  - it's very opaque what is tested and what not by the unit tests. This will make changing the code in future a rather convoluted task as changing private functions might break mock unit tests that depend on them. We have a coding guideline of testing only exported functions, and this is doing the opposite
  - there is no unit test coverage at all for a lot of the code defined in the actual implementations

The structure I would expect for this is something like a mock client, that implements only functions that already exists in the original [client](https://github.com/spiffe/go-spiffe/blob/main/v2/workloadapi/client.go), something like dial, close, get resource, list resources, create resource, delete resource. The `ControllerAPIClient` and `EntrypointerAPIClient` would only be responsible for implementing higher abstraction functions like  `VerifyTaskRunResults` and `Sign`. They would take a client as input, which would be the actual client at runtime, or the mock client at unit test time.

/cc @tektoncd/core-maintainers thoughts?

pxp928 commented on pkg/spire/entrypointer.go:
```diff
@@ -0,0 +1,78 @@
+	client *workloadapi.Client
+}
+
+func (w *spireEntrypointerAPIClient) checkClient(ctx context.Context) error {
```
Changed to `setupClient` to better reflect the function.

pxp928 commented:
/kind feature

wlynch commented on pkg/spire/verify.go:
```diff
@@ -0,0 +1,234 @@
+	return verifySignature(pub, sigValue, resultValue)
+}
+
+func verifySignature(pub interface{}, signature string, value string) error {
```
This doesn't make a functional difference, but should we use this for better type hinting from the func args? (also for verifyResult)

```suggestion
func verifySignature(pub crypto.PublicKey, signature string, value string) error {
```

pxp928 replied:
Agreed. Changed.

vdemeester commented on pkg/spire/config/config.go:
```diff
@@ -0,0 +1,68 @@
+	}
+	if len(unset) > 0 {
+		sort.Strings(unset)
+		return fmt.Errorf("found unset image flags: %s", unset)
```
I don't think the error is correct here (no image involevd)

pritidesai replied:
I think it will be great to make this error more descriptive. This error will be met when `SpireConfig` is missing all four required flags which in turn means no image initialized 🤔

pxp928 replied:
changed to "found unset spire configuration flags"

vdemeester reviewed (commented):
Agreeing with @wlynch and @afrittoli, in term of code organization especially for test.

pxp928 commented on pkg/spire/controller.go:
```diff
@@ -0,0 +1,276 @@
+	"fmt"
+	"time"
+
+	entryv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/server/entry/v1"
+	spiffetypes "github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+
+	corev1 "k8s.io/api/core/v1"
+
```
No reason. Accidental. Fixed.

pxp928 commented on pkg/spire/entrypointer.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+func (w *spireEntrypointerAPIClient) getxsvid(ctx context.Context) *x509svid.SVID {
+	backoffSeconds := 2
+	var xsvid *x509svid.SVID
+	var err error
+	for i := 0; i < 20; i += backoffSeconds {
+		xsvid, err = w.client.FetchX509SVID(ctx)
+		if err == nil {
+			break
+		}
+		time.Sleep(time.Duration(backoffSeconds) * time.Second)
+	}
+	return xsvid
```
For this we are checking fetching the SVID every few seconds. The change suggested with break the logic. I changed the variable name to `backoff` as it might be the causing confusion.

wlynch replied:
My bad - you're 100% right! Ignore this.

pxp928 commented on pkg/spire/sign.go:
```diff
@@ -0,0 +1,107 @@
+
+	output := []v1beta1.PipelineResourceResult{}
+	p := pem.EncodeToMemory(&pem.Block{
+		Bytes: xsvid.Certificates[0].Raw,
```
Agreed. Fixed.

wlynch replied:
nit: I'd move the len check to here - it's not immediately obvious that it's being checked in getWorkloadSVID.

pxp928 commented on pkg/spire/spire.go:
```diff
@@ -0,0 +1,52 @@
+// ControllerAPIClient interface maps to the spire controller API to interact with spire
+type ControllerAPIClient interface {
+	Close() error
+	CreateEntries(ctx context.Context, tr *v1beta1.TaskRun, pod *corev1.Pod, ttl int) error
```
This TTL is for the how long the entries that are registered into the Spire Sever are valid. Currently we are also using the default timeout setup by the config when initializing it and it will be deleted when the `tr.isDone()`. 

```
// TTL for the entry is in seconds
ttl := config.FromContextOrDefaults(ctx).Defaults.DefaultTimeoutMinutes * 60
```

wlynch replied:
My thinking in making it a duration is that we just make the above line

```
ttl := config.FromContextOrDefaults(ctx).Defaults.DefaultTimeoutMinutes * time.Minute
```

instead of converting to seconds. wdyt?

lumjjb replied:
it does end up being translated back to `int` when talking to the SPIRE API, but am okay with either.

pxp928 replied:
Changed to `time.Duration` but like Brandon said it will have to translate back to `int` for the SPIRE API.

pxp928 commented on pkg/spire/verify.go:
```diff
@@ -0,0 +1,234 @@
+		return nil, err
+	}
+	trustPool := x509.NewCertPool()
+	for _, c := range x509Bundle[0].X509Authorities() {
```
A trust bundle is a collection of one or more certificate authority (CA) root certificates that the workload should consider trustworthy. Added the `len(x509Bundle)` check.

wlynch replied:
I could have phrased that better - I meant should we looping over `x509Bundle` (as opposed to just the authorities in first one) i.e.

```go
trustPool := x509.NewCertPool()
for _, bundle := range x509Bundle {
	for _, c := range bundle.X509Authorities() {
		trustPool.AddCert(c)
	}
	return trustPool, nil
}
```

pxp928 replied:
Yup. Added.

pxp928 commented:
/retry

pxp928 commented:
/retest

lumjjb commented on pkg/spire/spire_mock.go:
```diff
@@ -0,0 +1,209 @@
+	corev1 "k8s.io/api/core/v1"
+)
+
+// MockClient is a client used for mocking the this package for unit testing
+// other tekton components that use the spire entrypointer or controller client.
+//
```
Hi! Sorry about being late to reply. COVID happened :|

Wanted to share a bit more about the thought process around the mocking,

---
### For unit testing

I think the client mocking will work perfectly with unit tests.

----

### For integration

We explored several different approaches when trying to mock SPIRE, a lot of the difficulty in doing so was that test that use mocked SPIRE resulted in two separate processes interacting with the shared state for integration tests. Mocking the workload API for a single memory process is ok, but when two processes (in this case the entrypointer and the controller) need to be able to retrieve compatible crypto material, is where things get a bit messy since we need to maintain state across processes and isolated containers. We would have to mock a daemon server, which pretty much re-implements a bundled version of SPIRE.

The workaround was performing mocking at a slightly higher layer, which takes away the need to have shared state between 2 processes with a fixed seed which would result in the need of an external daemon.

I think there's a desire to have the mocked SPIRE daemon with compatible fake workload client, but it is something that will be at least a couple months work of a dedicated developer.

Another possibility here could also be to create determinism within the ecdsa/rsa crypto library, it's not something I've explored too much.

---

What are some thoughts here? Maybe client mocking for unit tests but keep the higher level mocking for now for the integration/e2e tests - until the SPIRE community makes more progress on a mock server?

wlynch replied:
No worries! Hope you're feeling better! 🙏 

That sounds good to me! My comment was more about which clients we are mocking for unit tests so that we cover more codepaths for the real implementation (in particular, I'm worried we might be missing stuff in sign.go / controller.go because spire_mock.go is using its own impl). I'd be okay with an additional higher level mock for e2e/integration tests if needed.

We may also want e2e tests that set up real spire, but I'm fine with making that out of scope for this PR.

pxp928 replied:
The e2e/integration tests already do set up the real spire. We have all this completed in the following two PRs (Phase 1 and Phase 2): https://github.com/tektoncd/pipeline/pull/4759 and https://github.com/tektoncd/pipeline/pull/4828

We are only using the mocked spire for unit testing with the various other packages (for example the pod and status).

lumjjb replied:
Oh sorry on that last "e2e" left that in there, meant to be just integration*. 

Ok so @wlynch so sounds like the plan is that we will add in the fakeworkload client for unit testing, for additional coverage and leave the alternative mocking (and use the injection mechanism instead) for integration tests? Does that sound about right?

wlynch replied:
Sounds great to me - thanks!

afrittoli replied:
I definitely would be happy to have mocks on the workload API for unit test.
For E2E tests I would tend to say we should use the full server instead of a mock.

pxp928 commented on pkg/spire/spire.go:
```diff
@@ -0,0 +1,52 @@
+	VerifyTaskRunResults(ctx context.Context, prs []v1beta1.PipelineResourceResult, tr *v1beta1.TaskRun) error
+}
+
+// EntrypointerAPIClient interface maps to the spire entrypointer API to interact with spire
```
same question as above.

pxp928 replied:
Added the links

pxp928 commented:
@vdemeester @afrittoli @wlynch Should be ready for another review. @lumjjb added the injection piece and I have added the `fakeworkloadapi` unit tests to test the spire "sign" and "verify" functionality. I have also added some missing pieces to the spire package that are needed for phase 2 (signed TaskRun). This will allow the other two follow on PRs to utilize all the functionality from the spire package for TaskRun results and status signing and verification.

Once the review is complete. I will squash all the commits into one before it gets merged.

pxp928 commented:
/restest

pxp928 commented:
[@vdemeester](https://github.com/vdemeester) [@afrittoli](https://github.com/afrittoli) [@wlynch](https://github.com/wlynch) Should be ready for another review. [@lumjjb](https://github.com/lumjjb) added the injection piece and I have added the fakeworkloadapi unit tests to test the spire "sign" and "verify" functionality. I have also added some missing pieces to the spire package that are needed for phase 2 (signed TaskRun). This will allow the other two follow on PRs to utilize all the functionality from the spire package for TaskRun results and status signing and verification.

Once the review is complete. I will squash all the commits into one before it gets merged.

pxp928 commented:
/retest

pxp928 commented:
/retest

pxp928 commented on .golangci.yml:
```diff
@@ -44,5 +44,6 @@ run:
   skip-dirs:
   - vendor
   - pkg/client
+  - pkg/spire/test
```
The spire test package had a bunch of linting errors but as it was copied over from spire, I was not sure if I made sense to fix it all.

wlynch replied:
Is this a verbatim copy, or are we making modifications?

(if we're making modifications, I'd prefer to keep linting on for our changes if possible)

we should also add instructions for how this was imported for if/when we need to update.

pritidesai replied:
If the test client needs to be part of the pipeline code base, we will need license headers in each file under `pkg/spire/test`, fix linting errors, and some kind of documentation on why ...

pxp928 replied:
It is verbatim copy so do we need liniting?

pxp928 replied:
Created an issue #5281 to track this.

pxp928 commented:
@afrittoli Also seems like we need your help to figure out how to initialize the context for client injection during initialization of the entrypoint, controller (cmd/entrypoint/main.go) etc.

pxp928 commented on pkg/spire/spire.go:
```diff
@@ -0,0 +1,52 @@
+	VolumeMountPath = "/spiffe-workload-api"
+)
+
+// ControllerAPIClient interface maps to the spire controller API to interact with spire
```
Added the links

pxp928 commented:
/retest

pxp928 commented:
/retest

afrittoli commented:
/assign @vdemeester @jerop @pritidesai - assigning TEP approvers to the matching PR review :)

wlynch commented on pkg/spire/test/fakeworkloadapi/workload_api_windows.go:
```diff
@@ -0,0 +1,65 @@
+//go:build windows
+// +build windows
```
Can we add a short blurb on why a windows version is required? e.g. do we need to use the winio pipe, or can we simplify this with basic tcp on localhost?

pxp928 replied:
This is the remanent of the copy of the fakeworkloadapi. I will remove it.

wlynch commented on pkg/spire/sign.go:
```diff
@@ -0,0 +1,148 @@
+
+	// Store Controller SVID
+	p := pem.EncodeToMemory(&pem.Block{
+		Bytes: xsvid.Certificates[0].Raw,
```
This could possibly also be len == 0? (probably worth having a sanity check here just in case).

pxp928 replied:
Added check:
```
	if len(xsvid.Certificates) == 0 {
		return errors.New("returned controller svid does not have certificates")
	}
```

wlynch commented on pkg/spire/spire.go:
```diff
@@ -0,0 +1,75 @@
+	// controllerSvidAnnotation TaskRun status annotation controller SVID Key
+	controllerSvidAnnotation = "tekton.dev/controller-svid"
+	// NotVerifiedAnnotation TaskRun status annotation not verified by spire key that get set when status match fails
+	NotVerifiedAnnotation = "tekton.dev/not-verified"
```
```suggestion
	VerifiedAnnotation = "tekton.dev/spire-verified"
```

A general rule of thumb I like to follow is avoiding double negatives. It helps avoid ambiguous cases like does the absence of this annotation mean that it was successfully verified?

Also recommend using `spire-verified` so this doesn't end up conflicting with other verification checks.

pxp928 replied:
Agreed. Changed to suggested.

wlynch reviewed (commented):
Looking good! Definitely digging this structure much more. 😎

pxp928 commented:
> Is this a verbatim copy, or are we making modifications?

Yes the `pkg/spire/test` folder is a verbatim copy from the upstream project. No modification were made.

lumjjb commented on pkg/spire/spire_mock.go:
```diff
@@ -0,0 +1,314 @@
+//
+// The MockClient implements both SpireControllerApiClient and SpireEntrypointerApiClient
+// and in addition to that provides the helper functions to define and query internal state.
+type MockClient struct {
```
Unfortunately the `MockClient` uses a lot of the `pkg/spire` internal helper functions. We could add a comment that this struct should not be used outside `tektoncd` testing. As another measure, we could also add a build tag for integration `// +build integration`. wdyt?

pritidesai commented on pkg/reconciler/testing/logger.go:
```diff
@@ -31,6 +31,16 @@ func SetupFakeContext(t *testing.T) (context.Context, []controller.Informer) {
 	return WithLogger(ctx, t), informer
 }
 
+// SetupDefaultContext sets up the the Context and the default filtered informers for the tests.
```
NIT: extra `the` in `sets up the the Context`

pxp928 replied:
fixed

pritidesai commented on pkg/reconciler/testing/logger.go:
```diff
@@ -51,3 +61,13 @@ func setupFakeContextWithLabelKey(t zaptest.TestingT) (context.Context, context.
 	return ctx, c, is
 }
+
+// setupDefaultContextWithLabelKey sets up the the Context and the default informers for the tests
```
NIT: extra `the` in `sets up the the Context`

pxp928 replied:
fixed.

pritidesai commented on pkg/reconciler/testing/logger.go:
```diff
@@ -51,3 +61,13 @@ func setupFakeContextWithLabelKey(t zaptest.TestingT) (context.Context, context.
+
+// setupDefaultContextWithLabelKey sets up the the Context and the default informers for the tests
+// The provided context includes the FilteredInformerFactory LabelKey.
+func setupDefaultContextWithLabelKey(t zaptest.TestingT) (context.Context, context.CancelFunc, []controller.Informer) {
```
just wondering how `setupDefaultContextWithLabelKey` is different from `setupFakeContextWithLabelKey`? it looks same 🤔 

https://github.com/tektoncd/pipeline/blob/00b14d78ad73c055451f2a714387cd03f6222b89/pkg/reconciler/testing/logger.go#L47-L53

pxp928 replied:
Clean up how the functions are called but the difference is the  `injection.Fake` to `injection.Default`. This is needed so that the fakeworkloadAPI can be called for unit testing the spire api within `spire_test.go`.

pritidesai commented on pkg/reconciler/testing/logger.go:
```diff
@@ -31,6 +31,16 @@ func SetupFakeContext(t *testing.T) (context.Context, []controller.Informer) {
 }
 
+// SetupDefaultContext sets up the the Context and the default filtered informers for the tests.
+func SetupDefaultContext(t *testing.T) (context.Context, []controller.Informer) {
```
same as `SetupFakeContext`:

https://github.com/tektoncd/pipeline/blob/00b14d78ad73c055451f2a714387cd03f6222b89/pkg/reconciler/testing/logger.go#L25-L32

pxp928 replied:
Removed `cloudEventClientBehaviour` as not needed but this is to call the above `setupDefaultContextWithLabelKey` used to test spire with the fakeworkloadAPI.

pxp928 commented on pkg/spire/spire_test.go:
```diff
@@ -0,0 +1,731 @@
+	cfg := &config.SpireConfig{}
+	cfg.SocketPath = wl.Addr()
+	cfg.TrustDomain = trustDomain
+	spireControllerClient := GetControllerAPIClient(ctx)
+	spireControllerClient.SetConfig(*cfg)
+
+	logger := logging.FromContext(ctx)
+
+	var (
+		cc = spireControllerClient
+	)
+	defer cc.Close()
```
Fixed.

pxp928 commented:
/retest

pxp928 commented:
/retest

pxp928 commented:
/retest

lumjjb commented:
@wlynch @pritidesai changes made PTAL!

wlynch reviewed (commented):
I think this is pretty much there! /lgtm

pxp928 commented on pkg/spire/test/fakeworkloadapi/workload_api_posix.go:
```diff
@@ -0,0 +1,17 @@
+//go:build !windows
```
removed.

pxp928 commented:
/retest

lumjjb commented:
/retest

lumjjb commented:
Hi @pritidesai we've made your requested! Can you take a look at the PR again - would like to get this across the line! Thanks!

wlynch reviewed (approved):
/lgtm

@pritidesai can you take another look at this for approval? 👀

vdemeester reviewed (approved):
/hold
@pritidesai @afrittoli one last review before removing the hold 🙏🏼

wlynch commented:
Since this PR already has an LGTM from me and has been sitting without additional feedback for weeks, I'm going to go ahead and remove the hold to let this submit.

Nothing in this PR is a one-way door decision and is not wired up to the controller yet, so there's not much risk in submitting this even if there is additional feedback. Feel free to continue to leave feedback on this PR - we can follow up with additional PRs if needed!

/remove-hold

afrittoli commented:
Great to see this moving forward! I'm sorry I wasn't able to re-review before, but I'm glad to see progress.
